### PR TITLE
[Utility] MOBA-like Shield

### PR DESCRIPTION
**Pull request by PruBlu**

### CODE REVIEW

File: src/battle_game/core/units/doom_tank.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  DoomTank(GameCore *game_core, uint32_t id, uint32_t player_id);
+  void Render() override;
+  void Update() override;
+
+ protected:
+  void RandomDo();
+  void Doomsday();
+  void DestroyAll();
+  void DoomTankUpdate();
+  void Leech_();
```
Comment (qinchuanhui) [inline src/battle_game/core/units/doom_tank.h:17]: Naming, `Leech`is ok

File: src/battle_game/core/units/doom_tank.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  uint32_t id_{0};
+  bool IsHit(glm::vec2 position) const;
+  bool IsLeeched = false;
+  bool IsShield = false;
+  uint32_t doomsday_count_down_{0};
+  uint32_t leech_count_down_{0};
+  uint32_t leech_time_{240};
+  uint32_t shield_count_down_{0};
+  uint32_t shield_time_{180};
+  double InitialHealth;
```
Comment (qinchuanhui) [inline src/battle_game/core/units/doom_tank.h:32]: naming issue

File: src/battle_game/core/obstacles/crater.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+         float rotation = 0.0f,
+         glm::vec2 scale = glm::vec2{1.0f, 1.0f});
+  void Render() override;
+  void Update() override;
+
+ private:
+  void Recover();
+  uint32_t crater_count_down_{180};
+  uint32_t recovery_count_down_{180};
+  bool isRecovered = false;
```
Comment (qinchuanhui) [inline src/battle_game/core/obstacles/crater.h:20]: Naming

**Review by qinchuanhui — CHANGES_REQUESTED** [review]

**Comment by qinchuanhui** [comment]
Please change your title and add some description through comments, both of which should introduce the work you finished

**Comment by PruBlu** [comment]
### [Utility] MOBA-like shield

- Feature:

Shield is a passive trait. while getting a health loss more than 5%, a shield of 0.8 of your health loss will be acquired. After 3 seconds it will automatically remove the rest of the shield. If the shield has been used up, nothing will happen.

Specially, it can work well with Leech skill, so the health recovery will be awfully awesome.

**Review by qinchuanhui — APPROVED** [review]

**Pull request closed without merging**